### PR TITLE
feature: rBasis unstaking -> Basis in rBasis collateral liquidation

### PR DESCRIPTION
**Pull request by cryptoboole**
This pull request adds the code and tests required to unstake rBasis when liquidated. 

**Todo**
- [ ] add documentation
- [ ] ensure there is more abstraction in `liquidate.ts`
- [ ] test with live market and adapt

### CODE REVIEW

File: src/libs/swaps/basis/rBasisSwap.ts [hunk 1]
```diff
@@ -0,0 +1,613 @@
...
+export async function getTokenAccount(connection, publicKey) {
+  const result = await connection.getAccountInfo(publicKey);
+  if (!result) return false;
+  const data = Buffer.from(result.data);
+  const account = parseTokenAccountData(publicKey, data);
+  return {
+    publicKey,
+    account,
+  };
+}
```
Comment (0xCactus) [inline src/libs/swaps/basis/rBasisSwap.ts:85]: can we move parseTokenAccountData and getTokenAccount to src/libs/utils.ts. Seem like handy methods

File: src/libs/swaps/basis/rBasisSwap.ts [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+    { pubkey: new PublicKey(PROGRAM_BASIS_STAKING_VAULT), isSigner: false, isWritable: true },
+    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
+  ];
+
+  return new TransactionInstruction({
+    keys,
+    programId: new PublicKey(PROGRAM_BASIS_STAKING),
+    data: Buffer.concat([sighash(SIGHASH_GLOBAL_NAMESPACE, StakingInstructionNames[StakingInstruction.unstake]), data]),
+  });
+};
```
Comment (0xCactus) [inline src/libs/swaps/basis/rBasisSwap.ts:130]: can we move this into src/libs/models/instructions/rbasis/unstake.ts

File: src/libs/swaps/basis/rBasisSwap.ts [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+
+export const unstakeBasis = async (
+  connection: Connection,
+  payer: Account,
+) => {
+  const ixs: TransactionInstruction[] = [];
+
+  // get associated token account for Basis (or create if doens't exist)
+  const BasisAccount = await Token.getAssociatedTokenAddress(
```
Comment (0xCactus) [inline src/libs/swaps/basis/rBasisSwap.ts:139]: can save compute by moving this to after we check for rbasis amount in wallet and only fetch/create ata when needed

File: src/libs/swaps/basis/rBasisSwap.ts [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+  // get associated token account for rBasis
+  const rBasisAccount = await Token.getAssociatedTokenAddress(
+    ASSOCIATED_TOKEN_PROGRAM_ID,
+    TOKEN_PROGRAM_ID,
+    new PublicKey(MINT_RBASIS),
+    payer.publicKey,
+  );
+  const rBasisAccountInfo = await connection.getAccountInfo(rBasisAccount);
+  const rBasisTokenAccount = await getTokenAccount(connection, rBasisAccount);
+  if (!rBasisAccountInfo || !rBasisTokenAccount) throw new Error('this function requires active rBasis associated token account');
```
Comment (0xCactus) [inline src/libs/swaps/basis/rBasisSwap.ts:169]: instead of throwing an error, we can just return. Not all wallets will have an rbasis ata

**Review by 0xCactus — COMMENTED** [review]
Looks good overall. Left a few minor comments

File: src/libs/swaps/basis/rBasisSwap.ts (hunk 2 above)
Comment (cryptoboole) [reply src/libs/swaps/basis/rBasisSwap.ts:130]: 👍

File: src/libs/swaps/basis/rBasisSwap.ts (hunk 4 above)
Comment (cryptoboole) [reply src/libs/swaps/basis/rBasisSwap.ts:169]: 👍

**Comment by cryptoboole** [comment]
Will make some adjustments and submit another commit for your review @0xCactus

**Comment by 0xCactus** [comment]
included rBasis unstaking change in https://github.com/solendprotocol/liquidator/pull/44

**Pull request closed without merging**